### PR TITLE
Fix the gradient computation of dynamic stitch.

### PR DESCRIPTION
**Pull request by codrut3**
Currently the gradient of tf.dynamic_stitch is not correct for duplicated indices. This is
issue #7397. @drasmuss submitted a pull request for this issue, but it was ultimately not merged
due to a performance drop.

While working on this, I realised that the problem is more complicated than what follows
from the discussion in #7487. In fact, the solution proposed in #7487 is not correct, because
one can have duplicated indices in the same input tensor. The following example shows this:

```
import tensorflow as tf

x = tf.zeros((3,))
y = tf.dynamic_stitch([[0, 1, 0]], [x])

with tf.Session() as sess:
    print("y")
    print(sess.run(y))

    analytic, numeric = tf.test.compute_gradient(x, (3,), y, (2,))
    print("analytic")
    print(analytic)
    print("numeric")
    print(numeric)
```

The output (even with #7487) is

```
y
[ 0.  0.]
analytic
[[ 1.  0.]
 [ 0.  1.]
 [ 1.  0.]]
numeric
[[ 0.          0.        ]
 [ 0.          0.99998707]
 [ 0.99998707  0.        ]]
```

My fix is to add a new kernel op `GatherDisjoint`, that can handle all the inputs together.
This replaces the `n` calls to `gather` in `data_flow_grad._DynamicStitchGrads` with a
single `gather_disjoint`.

The implementation is very similar to `gather`, I just zero out duplicated slices at the end.
I've put this op in core/, if you would like I can move it to contrib/, just let me know where.

I've also added tests for the op and new tests for the gradient of dynamic stitch. Here
I inspired myself from @drasmuss pull request.

I've run the script by @drasmuss again:

```
import time

import numpy as np
import tensorflow as tf

stitch_size = 1000
n_inputs = 10
input_shape = (100, 50, 50)
reps = 10

with tf.device("cpu:0"):
    idxs = [tf.constant(np.random.randint(stitch_size, size=input_shape[0]), dtype=tf.int32)
            for _ in range(n_inputs)]
    vals = [tf.constant(np.random.uniform(-1, 1, size=input_shape), dtype=tf.float32)
            for _ in range(n_inputs)]
    y = tf.dynamic_stitch(idxs, vals)
    grad = tf.gradients(y, vals)

total_time = 0.0
for _ in range(reps):
    with tf.Session() as sess:
        start = time.time()
        sess.run(grad)
        total_time += time.time() - start

print(total_time / reps)
```

The results are as follows:
with my fix:
CPU - 0.09455678462982178
GPU - 0.09739606380462647
without my fix:
CPU - 0.11728813648223876
GPU - 0.11805038452148438

So with the fix it runs faster. On GPU the slowdown is due to the memory copies from host to device
and back, for input and output (I've checked with nvprof), and not because of my fix.

I've run this script using local builds of the same master branch, with and without my patch.

Let me know if the op should be moved somewhere else, or it should be made hidden.

### CODE REVIEW

**Comment by caisq** [comment]
@tensorflow-jenkins test this please

**Comment by codrut3** [comment]
The reasons for failure are:
- dynamic stitch was modified since I started working on this and my commit does not work properly with the new modifications
- the API changed and the goldens should be updated (as I've said, I can move the op to contrib)
- for the GPU, XLA and Windows there is a strange failure involving a hudson plugin that I don't understand

I will fetch the latest version of tensorflow and commit an update.

**Comment by caisq** [comment]
ping @aselle

**Comment by codrut3** [comment]
Would you mind testing the PR? I think all tests should pass.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by codrut3** [comment]
Hi @martinwicke ,

I fixed the failure, would you mind testing again?

**Comment by martinwicke** [comment]
ping @aselle for review.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 14 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 16 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 157 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 172 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 187 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: It has been 202 days with no activity and the `awaiting review` label was assigned. Can you please take a look?

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 217 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 232 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 247 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 262 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 277 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 292 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 307 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 322 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 337 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 352 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 367 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @aselle: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/core/kernels/gather_disjoint_op_test.cc [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+  random::PhiloxRandom philox(301, 17);
+  random::SimplePhilox rnd(&philox);
+  std::vector<Index> indices_vec;
+  indices_vec.reserve(kLookups);
+  for (int i = 0; i < kLookups; i++) {
+    indices_vec.push_back(rnd.Uniform(kRows));
+  }
+  Tensor indices(DataTypeToEnum<Index>::value, TensorShape({kLookups}));
+  for (int i = 0; i < indices_vec.size(); i++) {
```
Comment (shahzadlone) [inline tensorflow/core/kernels/gather_disjoint_op_test.cc]: Prefer pre-increment over post-increment.

```diff
+  for (int i = 0; i < indices_vec.size(); ++i) {
-   for (int i = 0; i < indices_vec.size(); i++) {
```

**Comment by rthadur** [comment]
@codrut3 wondering if this PR valid , if yes can you please resolve conflicts.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by codrut3** [comment]
Hi Rajeshwar,

   I am sorry for taking so long to answer. I think the PR is still valid
but needs to be updated in view of recent changes. I'll have a look and
reopen it if that's the case.

   Best,
     Codrut

**Pull request closed without merging**